lib/compatibility.bzl

```diff
@@ -0,0 +1,185 @@
+
+    Returns:
+      A native `select()` which maps any of the settings to the incompatible target.
+    """
+    compat_name = " incompatible with " + _get_name_from_target_list(settings)
+    _maybe_make_unique_incompatible_value(compat_name)
+
+    return selects.with_or({
+        "//conditions:default": [],
+        tuple(settings): [":" + compat_name],
+    })
+
+def _any_of(settings):
+    """Create a `select()` for `target_compatible_with` which matches any of the given settings.
+
+    Any of the settings will resolve to an empty list, while the default condition will map to
+    an incompatible `constraint_value` for the purpose of target skipping.
+
+    In other words, use this function to make target incompatible unless one or more of the
+    settings are true.
+
+    ```python
+    cc_binary(
+        name = "bin",
+        srcs = ["bin.cc"],
+        # This target can only be built for Linux or Mac.
+        target_compatible_with = compatibility.any_of([
+            "@platforms//os:linux",
+            "@platforms//os:macos",
+        ]),
+    )
+    ```
+
+    Args:
+      settings: A list of `config_settings` or `constraint_value` targets.
+
+    Returns:
+      A native `select()` which maps any of the settings an empty list.
+    """
+    compat_name = " compatible with any of " + _get_name_from_target_list(settings)
+    _maybe_make_unique_incompatible_value(compat_name)
+
+    return selects.with_or({
+        tuple(settings): [],
+        "//conditions:default": [":" + compat_name],
+    })
+
+def _all_of(settings):
+    """Create a `select()` for `target_compatible_with` which matches all of the given settings.
+
+    All of the settings must be true to get an empty list. Failure to match will result
+    in an incompatible `constraint_value` for the purpose of target skipping.
+
+    In other words, use this function to make a target incompatible unless all of the settings are
+    true.
+
+    Example:
+
+    ```python
+    config_setting(
+        name = "dbg",
+        values = {"compilation_mode": "dbg"},
+    )
+
+    cc_binary(
+        name = "bin",
+        srcs = ["bin.cc"],
+        # This target can only be built for Linux in debug mode.
+        target_compatible_with = compatibility.all_of([
+            ":dbg",
+            "@platforms//os:linux",
+        ]),
+    )
+    ```
+
+    See also: `selects.config_setting_group(match_all)`
+
+    Args:
+      settings: A list of `config_setting` or `constraint_value` targets.
+
+    Returns:
+      A native `select()` which is "incompatible" unless all settings are true.
+    """
+    group_name = _get_name_from_target_list(settings, joiner = " and ")
+    compat_name = " compatible with all of " + group_name
+    _maybe_make_unique_incompatible_value(compat_name)
+
+    # all_of can only be accomplished with a config_setting_group.match_all.
+    if not native.existing_rule(group_name):
```

Comment:
We'd like to get rid of `native.existing_rule` and we don't really support "anonymous" target definitions. Are there other ways how to implement this without using it? 

Maybe copying `selects.config_setting_group` pattern, that is add a target for `all_of` and then refer to it in the target_compatible_with.

Another alternative is to construct and return `select`s, without creating a target. I know this is possible for `match_any(a,b)` (by using `select{a:a,b:b}`) and it seems to be possible for `match_all` (just returning a list should do it, if I'm not mistaken). It might even work with some types of compositions.

Reply:
I think we could solve this by chaining `select()` statements, but only if we pre-define a pre-determined number of incompatible `constraint_value` targets. E.g. if we wanted to support up to 50 settings in `all_of`, we could do this:
```python
# lib/compatibility/BUILD
[constraint_setting(
    name = "all_of_" + str(i),
    constraint_value = "@platforms//:incompatible_setting",
) for i in range(50)]
```
Then `all_of` could be implemented roughly like so:
```python
def all_of(*settings):
    result = []
    for i, setting in enumerate(settings):
        result += select({
            setting: [],
            "//conditions:default": "@bazel_skylib//lib/compatibility:all_of_" + str(i),
        })
    return result
```
So if _any_ of the `select`s fail, then the target would be incompatible.

Am I thinking about this correctly?

Reply:
I took the liberty of implementing this. What do you think?